### PR TITLE
Add general purpose host memory allocator reference to cuIO with a demo of pooled-pinned allocation.

### PR DESCRIPTION
**Pull request by nvdbaranec**
This PR adds a new interface to cuIO which controls where host memory allocations come from. It adds two core functions:

Addresses https://github.com/rapidsai/cudf/issues/14314

```
rmm::host_async_resource_ref set_host_memory_resource(rmm::host_async_resource_ref mr);
rmm::host_async_resource_ref get_host_memory_resource();
```

`cudf::io::hostdevice_vector` was currently implemented in terms of a `thrust::host_vector<>` that explicitly uses an allocator called `pinned_host_vector`.  I copied that and made a new class called `rmm_host_vector` which takes any host_resource_ref.  This probably makes `pinned_host_vector` obsolete.  

Parquet benchmarks have a new commandline option which lets you toggle between 3 modes:

```
--cuio_host_mem pinned              (the default, an unpooled, pinned memory source)
--cuio_host_mem pinned_pool         (the pooled/pinned resource)
```

The ultimate intent here is to reduce the cpu-side overhead of the setup code that comes before the decode kernels in the parquet reader.  The wins are pretty significant for our faster kernels (that is, where we are less dominated by gpu time)

Edit: Updated to use newly minted resource ref types from rmm itself.  I also switched the type to be `host_async_resource_ref` even though in this case the user (`thrust::host_vector`) doesn't explicitly go through the async path.  In addition, the pageable memory path (an experimental feature) has been removed.

Pinned
```
| data_type |    io_type    | cardinality | run_length | Samples | CPU Time  | Noise | GPU Time  | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|-----------|---------------|-------------|------------|---------|-----------|-------|-----------|-------|------------------|-------------------|-------------------|
|  INTEGRAL | DEVICE_BUFFER |           0 |          1 |     25x | 20.443 ms | 0.45% | 20.438 ms | 0.45% |      26268890178 |         1.072 GiB |       498.123 MiB |
|  INTEGRAL | DEVICE_BUFFER |        1000 |          1 |     26x | 19.571 ms | 0.42% | 19.565 ms | 0.42% |      27440146729 |       756.210 MiB |       161.438 MiB |
|  INTEGRAL | DEVICE_BUFFER |           0 |         32 |     28x | 18.150 ms | 0.18% | 18.145 ms | 0.18% |      29587789525 |       602.424 MiB |        27.720 MiB |
|  INTEGRAL | DEVICE_BUFFER |        1000 |         32 |     29x | 17.306 ms | 0.37% | 17.300 ms | 0.37% |      31032523423 |       597.181 MiB |        14.403 MiB |
```

Pooled/pinned
```
| data_type |    io_type    | cardinality | run_length | Samples | CPU Time  | Noise | GPU Time  | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|-----------|---------------|-------------|------------|---------|-----------|-------|-----------|-------|------------------|-------------------|-------------------|
|  INTEGRAL | DEVICE_BUFFER |           0 |          1 |    117x | 17.258 ms | 0.50% | 17.254 ms | 0.50% |      31115706389 |         1.072 GiB |       498.123 MiB |
|  INTEGRAL | DEVICE_BUFFER |        1000 |          1 |     31x | 16.413 ms | 0.43% | 16.408 ms | 0.43% |      32719609450 |       756.210 MiB |       161.438 MiB |
|  INTEGRAL | DEVICE_BUFFER |           0 |         32 |    576x | 14.885 ms | 0.58% | 14.881 ms | 0.58% |      36077859564 |       602.519 MiB |        27.720 MiB |
|  INTEGRAL | DEVICE_BUFFER |        1000 |         32 |     36x | 14.069 ms | 0.48% | 14.065 ms | 0.48% |      38171646940 |       597.243 MiB |        14.403 MiB |
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by GregoryKimball** [comment]
Nice work @nvdbaranec !!

**Comment by hyperbolic2346** [comment]
I like this ability. My only question is if we should follow the current optional memory resource passed into functions or if we should add this as a set/get.

```
table_with_metadata read_parquet(
  parquet_reader_options const& options,
  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource());
```
Maybe this becomes:
```
table_with_metadata read_parquet(
  parquet_reader_options const& options,
  rmm::cuda_stream_view stream        = cudf::get_default_stream(),
  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource(),
  cudf::host_resource_ref* host_mr    = rmm::mr::pinned_memory_resource);
```

I don't know where all this applies and the trouble of passing it through.

File: cpp/include/cudf/utilities/resource_ref.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+#pragma once
+
+#include <cuda/memory_resource>
+
+namespace cudf {
+
+using host_resource_ref       = cuda::mr::resource_ref<cuda::mr::host_accessible>;
+using async_host_resource_ref = cuda::mr::async_resource_ref<cuda::mr::host_accessible>;
```
Comment (harrism) [inline cpp/include/cudf/utilities/resource_ref.hpp:24]: Can you please open an RMM PR to add these aliases to RMM's `resource_ref.hpp`, and then use them instead of keeping them to libcudf?

File: cpp/benchmarks/fixture/nvbench_fixture.hpp [hunk 2]
```diff
@@ -25,12 +26,18 @@
 #include <rmm/mr/device/owning_wrapper.hpp>
 #include <rmm/mr/device/per_device_resource.hpp>
 #include <rmm/mr/device/pool_memory_resource.hpp>
+#include <rmm/mr/host/host_memory_resource.hpp>
+#include <rmm/mr/host/new_delete_resource.hpp>
+#include <rmm/mr/host/pinned_memory_resource.hpp>
```
Comment (harrism) [inline cpp/benchmarks/fixture/nvbench_fixture.hpp:31]: This is the old, to be deprecated pinned memory resource. Please use `<rmm/mr/pinned_host_memory_resource.hpp>

File: cpp/benchmarks/fixture/nvbench_fixture.hpp [hunk 3]
```diff
@@ -75,23 +82,64 @@ struct nvbench_base_fixture {
...
+    static std::shared_ptr<rmm::mr::pinned_memory_resource> mr =
+      std::make_shared<rmm::mr::pinned_memory_resource>();
+    return host_resource_ref(*mr);
+  }
+
+  inline auto make_cuio_host_pinned_pool()
+  {
+    using host_pooled_mr = rmm::mr::pool_memory_resource<rmm::mr::pinned_memory_resource>;
+    static std::shared_ptr<host_pooled_mr> mr = std::make_shared<host_pooled_mr>(
+      std::make_shared<rmm::mr::pinned_memory_resource>().get(), size_t{1} * 1024 * 1024 * 1024);
```
Comment (harrism) [inline cpp/benchmarks/fixture/nvbench_fixture.hpp:96]: I can't even believe this works. Please use the new `rmm::mr::pinned_host_memory_resource`, not the old (`host_memory_resource`) one.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+   *  @param x The other \p rmm_host_allocator of interest.
+   *  @return This method always returns \c false.
+   */
+  __host__ __device__ inline bool operator!=(rmm_host_allocator const& x) const
+  {
+    return !operator==(x);
+  }
+
+ private:
+  cudf::host_resource_ref mr;
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:223]: I'm surprised you don't want stream-ordered allocation for your pinned memory. Without it the `pool_memory_resource` will do everything on the default stream.

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 5]
```diff
@@ -66,19 +61,11 @@ class hostdevice_vector {
...
-    }
-
-    std::visit(
-      [&](auto&& v) {
-        v.reserve(max_size);
-        v.resize(initial_size);
-        host_data = v.data();
-      },
-      h_data_owner);
+    h_data_owner = std::make_unique<cudf::detail::rmm_host_vector<T>>(
```
Comment (harrism) [inline cpp/src/io/utilities/hostdevice_vector.hpp:64]: Do you never want to access the pinned host memory from the device? Because your `cudf::host_resource_ref` only defines the `host_accessible` property, you are saying that the resource is only required to allocate host-accessible memory. But the pinned resource allocates memory that is accessible from host and device. So you could avoid memcopies when using it for host_device vectors...

**Review by harrism — CHANGES_REQUESTED** [review]
I'm really surprised the old host_memory_resource works with the pool. I added `rmm::mr::pinned_host_memory_resource` (which implements the `cuda::mr::async_memory_resource` and `cuda::mr::memory_resource` concepts instead of deriving from `host_memory_resource`) specifically to enable use with `pool_memory_resource`. Please use it instead of the old one.

File: cpp/include/cudf/utilities/resource_ref.hpp (hunk 1 above)
Comment (nvdbaranec) [reply cpp/include/cudf/utilities/resource_ref.hpp:24]: Done. Wasn't able to edit labels on the PR (non-breaking, cpp) though.

https://github.com/rapidsai/rmm/pull/1479

File: cpp/src/io/utilities/hostdevice_vector.hpp (hunk 5 above)
Comment (vuule) [reply cpp/src/io/utilities/hostdevice_vector.hpp:64]: At this point we don't directly access the pinned memory from the device, and don't plan to. We measured that explicit copies are faster in cuIO.
Having a resource that only requires host access allows the resource to fall back to pageable memory if it's unable to provide a pinned memory buffer. This is the current behavior of the resource that spark plugin uses so everything lines up well.

Comment (harrism) [reply cpp/src/io/utilities/hostdevice_vector.hpp:64]: Got it.

File: cpp/src/io/utilities/config_utils.cpp [hunk 6]
```diff
@@ -80,4 +87,36 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
...
+
+inline auto& host_mr()
+{
+  static rmm::host_async_resource_ref host_mr = default_pinned_mr();
+  return host_mr;
+}
+
+}  // namespace detail
+
+void set_current_host_memory_resource(rmm::host_async_resource_ref mr)
```
Comment (abellina) [inline cpp/src/io/utilities/config_utils.cpp:110]: It would be great if this returned the old resource on the set. It would follow the pattern that was set for the device resources here: https://github.com/rapidsai/rmm/blob/f132d4b0daa976e1ec6cbcef24f5454fe510a394/include/rmm/mr/device/per_device_resource.hpp#L246

Comment (abellina) [reply cpp/src/io/utilities/config_utils.cpp:110]: not a blocker.

File: cpp/src/io/utilities/config_utils.cpp [hunk 7]
```diff
@@ -80,4 +87,36 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
...
+  return map_lock;
+}
+
+inline rmm::host_async_resource_ref default_pinned_mr()
+{
+  static rmm::mr::pinned_host_memory_resource default_mr{};
+  return default_mr;
+}
+
+inline auto& host_mr()
```
Comment (jrhemstad) [inline cpp/src/io/utilities/config_utils.cpp:102]: Required: You need this. See https://github.com/rapidsai/rmm/issues/826 

Probably a good idea to wrap it in a macro. See https://github.com/rapidsai/rmm/blob/branch-24.04/include/rmm/detail/export.hpp 

```suggestion
__attribute__((visibility("default")))
inline auto& host_mr()
```

File: cpp/src/io/utilities/config_utils.cpp [hunk 8]
```diff
@@ -80,4 +87,36 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
...
+
+}  // namespace detail
+
+void set_current_host_memory_resource(rmm::host_async_resource_ref mr)
+{
+  std::lock_guard<std::mutex> lock{detail::host_mr_lock()};
+  detail::host_mr() = mr;
+}
+
+rmm::host_async_resource_ref get_current_host_memory_resource()
```
Comment (jrhemstad) [inline cpp/src/io/utilities/config_utils.cpp:116]: suggestion (non-blocking): You can drop the "current" if you want. I know this was based on `get_current_device_memory_resource()`, but there the "current" adjective describes the device, not the memory resource, i.e., "The memory resource on the _current device_." 

The closest equivalent for a host memory resource would be a NUMA node. If you wanted to be fancier, you could emulate the `per_device_memory_resource` logic and extend it to have `per_numa_node_memory_resource` as well as `get/set_nearest_numa_memory_resource()` that would query the NUMA node of the calling thread.

Or you could just skip the NUMA stuff all together.

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 9]
```diff
@@ -190,7 +177,7 @@ class hostdevice_vector {
   }
 
  private:
-  std::variant<thrust::host_vector<T>, cudf::detail::pinned_host_vector<T>> h_data_owner;
+  std::unique_ptr<cudf::detail::rmm_host_vector<T>> h_data_owner;
```
Comment (jrhemstad) [inline cpp/src/io/utilities/hostdevice_vector.hpp:180]: Any reason to not just have it be a plain data member? 
```suggestion
 cudf::detail::rmm_host_vector<T> h_data_owner;
```

**Review by vuule — COMMENTED** [review]
flush

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+  rmm_host_allocator& operator=(rmm_host_allocator const& col)
+  {
+    mr = col.mr;
+    return *this;
+  }
+
+  /**
+   * @brief rmm_host_allocator's null destructor does nothing.
+   */
+  __host__ __device__ inline ~rmm_host_allocator() {}
```
[the remark this answers is not in the document]
Comment (harrism) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:125]: Why would this class ever be destroyed in device code? For that matter why would it ever be created device code, you can't call `mr.allocate` in device code, so this allocator wouldn't be useful on the device.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+  __host__ __device__ inline pointer address(reference r) { return &r; }
+
+  /**
+   * @brief This method returns the address of a \c const_reference
+   *  of interest.
+   *
+   *  @param r The \c const_reference of interest.
+   *  @return \c r's address.
+   */
+  __host__ __device__ inline const_pointer address(const_reference r) { return &r; }
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:143]: What are these for, and why do they need to be members of this class? Why are these functions, the destructor, and the comparison ops `__host__ __device__`?

File: cpp/benchmarks/fixture/nvbench_fixture.hpp [hunk 12]
```diff
@@ -75,23 +81,56 @@ struct nvbench_base_fixture {
...
     std::cout << "RMM memory resource = " << rmm_mode << "\n";
+
+    cudf::io::set_current_host_memory_resource(create_cuio_host_memory_resource(cuio_host_mode));
+    std::cout << "CUIO host memory resource = " << cuio_host_mode << "\n";
   }
 
   std::shared_ptr<rmm::mr::device_memory_resource> mr;
   std::string rmm_mode{"pool"};
+
+  // host_resource_ref cuio_host_mr;
```
Comment (miscco) [inline cpp/benchmarks/fixture/nvbench_fixture.hpp:132]: Should that comment go now that we use `set_current_host_memory_resource` above?

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+  /**
+   * @brief Copy constructor
+   */
+  rmm_host_allocator(rmm_host_allocator const& other) : mr(other.mr) {}
+
+  /**
+   * @brief Move constructor
+   */
+  rmm_host_allocator(rmm_host_allocator&& other) : mr(std::move(other.mr)) {}
```
[the remark this answers is not in the document]
Comment (miscco) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:111]: We should default these. Due to the destructor we cannot omit them though

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (miscco) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:125]: Yeah its a bit off that this is the only SMF that is `__host__ __device__`

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  __host__ __device__ inline ~rmm_host_allocator() {}
+
+  /**
+   * @brief This method returns the address of a \c reference of
+   *  interest.
+   *
+   *  @param r The \c reference of interest.
+   *  @return \c r's address.
+   */
+  __host__ __device__ inline pointer address(reference r) { return &r; }
```
Comment (miscco) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:134]: If you need the address of a reference, you should use `cuda::std::addressof` this seems like a bad remnant from long ago

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+   *  @param cnt The number of objects to allocate.
+   *  @return a \c pointer to the newly allocated objects.
+   *  @note The second parameter to this function is meant as a
+   *        hint pointer to a nearby memory location, but is
+   *        not used by this allocator.
+   *  @note This method does not invoke \p value_type's constructor.
+   *        It is the responsibility of the caller to initialize the
+   *        objects at the returned \c pointer.
+   */
+  __host__ inline pointer allocate(size_type cnt, const_pointer /*hint*/ = 0)
```
Comment (miscco) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:157]: Should we omit the `__host__` as that is the default anyway?

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+   *  @param x The other \p rmm_host_allocator of interest.
+   *  @return This method always returns \c false.
+   */
+  __host__ __device__ inline bool operator!=(rmm_host_allocator const& x) const
+  {
+    return !operator==(x);
+  }
+
+ private:
+  rmm::host_async_resource_ref mr;
```
Comment (miscco) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:212]: I want to make sure that it is clear that this is a reference. So the user is responsible for keeping the resource alife outside of the allocator.

That might be a common source of error

File: cpp/src/io/utilities/config_utils.cpp [hunk 17]
```diff
@@ -80,4 +87,36 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
...
+{
+  std::lock_guard<std::mutex> lock{detail::host_mr_lock()};
```
Comment (miscco) [inline cpp/src/io/utilities/config_utils.cpp:118]: I believe we are in C++17 only, so you can drop the explicit template arguments in favor of CTAD:
```suggestion
  std::lock_guard lock{detail::host_mr_lock()};
```

Applies throughout

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:1]: Sounds good.  This is definitely worth replacing.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 11 above)
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:143]: Both of these classes are just a copy of `pinned_host_vector` and `pinned_allocator`, with some tweaks.  
https://github.com/rapidsai/cudf/blob/branch-24.04/cpp/include/cudf/detail/utilities/pinned_host_vector.hpp

I'll clean up the cruft.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 14 above)
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:134]: See this comment:
https://github.com/rapidsai/cudf/pull/15079#discussion_r1514808686

I didn't clean out the device-side stuff.  Doing that now.

File: cpp/src/io/utilities/config_utils.cpp (hunk 6 above)
Comment (nvdbaranec) [reply cpp/src/io/utilities/config_utils.cpp]: Done.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 16 above)
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:206]: Are you thinking comment on the member or something bigger in the description of the class itself?

File: cpp/include/cudf/utilities/export.hpp [hunk 18]
```diff
@@ -0,0 +1,26 @@
...
+#pragma once
+
+// Macros used for defining symbol visibility, only GLIBC is supported
+#if (defined(__GNUC__) && !defined(__MINGW32__) && !defined(__MINGW64__))
+#define CUDF_EXPORT __attribute__((visibility("default")))
+#define CUDF_HIDDEN __attribute__((visibility("hidden")))
+#else
+#define CUDF_EXPORT
+#define CUDF_HIDDEN
+#endif
```
Comment (vuule) [inline cpp/include/cudf/utilities/export.hpp:26]: pretty sure style check will fail without a newline here

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 11 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp]: This begs the question, why do those classes have `__host__ __device__` code?

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 19]
```diff
@@ -0,0 +1,225 @@
...
+  inline pointer address(reference r) { return std::addressof(r); }
+
+  /**
+   * @brief This method returns the address of a \c const_reference
+   *  of interest.
+   *
+   *  @param r The \c const_reference of interest.
+   *  @return \c r's address.
+   */
+  inline const_pointer address(const_reference r) { return std::addressof(r); }
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:148]: Still don't see why these functions are needed. The caller of these could just call `std::addressof` directly.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 20]
```diff
@@ -0,0 +1,225 @@
...
+   *        not used by this allocator.
+   *  @note This method does not invoke \p value_type's constructor.
+   *        It is the responsibility of the caller to initialize the
+   *        objects at the returned \c pointer.
+   */
+  inline pointer allocate(size_type cnt, const_pointer /*hint*/ = 0)
+  {
+    if (cnt > this->max_size()) { throw std::bad_alloc(); }  // end if
+    return static_cast<pointer>(
+      mr.allocate_async(cnt * sizeof(value_type), THRUST_MR_DEFAULT_ALIGNMENT, stream));
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:166]: I wonder if you should use `rmm::RMM_DEFAULT_HOST_ALIGNMENT` instead. They are defined to the same thing, but the RMM version has `HOST` in the name, and this class is called `rmm_host_allocator` so RMM is closer.

That is defined in `rmm/aligned.hpp`.

File: cpp/benchmarks/fixture/nvbench_main.cpp [hunk 21]
```diff
@@ -21,20 +21,22 @@
...
-    if (arg == cudf::detail::rmm_mode_param) { \
-      i += 2;                                  \
-    } else {                                   \
-      m_args.push_back(arg);                   \
-    }                                          \
-  }                                            \
+#define NVBENCH_MAIN_PARSE(argc, argv)                     \
+  nvbench::option_parser parser;                           \
+  std::vector<std::string> m_args;                         \
+  for (int i = 0; i < argc; ++i) {                         \
```
Comment (harrism) [inline cpp/benchmarks/fixture/nvbench_main.cpp:30]: This can be for a separate PR, but in cuDF tests (and other RAPIDs tests) we use the cxxopts library to parse command line arguments. Might be good to use it here too.  See also https://github.com/NVIDIA/nvbench/issues/87

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 22]
```diff
@@ -0,0 +1,207 @@
...
+
+  /**
+   * @brief Move constructor
+   */
+  rmm_host_allocator(rmm_host_allocator&& other) = default;
+
+  /**
+   * @brief Assignment operator
+   */
+  rmm_host_allocator& operator=(rmm_host_allocator const& col)
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:121]: Nit, non-blocking. I wouldn't call this `col`.

```suggestion
  rmm_host_allocator& operator=(rmm_host_allocator const& other)
```

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 23]
```diff
@@ -0,0 +1,207 @@
...
+  inline ~rmm_host_allocator() {}
+
+  /**
+   * @brief This method allocates storage for objects in host memory.
+   *
+   *  @param cnt The number of objects to allocate.
+   *  @return a \c pointer to the newly allocated objects.
+   *  @note The second parameter to this function is meant as a
+   *        hint pointer to a nearby memory location, but is
+   *        not used by this allocator.
```
Comment (harrism) [inline cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:139]: What is the "hint" about?

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 16 above)
Comment (harrism) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:197]: I would make it in the description of the class.

**Review by harrism — APPROVED** [review]
Great to see this getting close.  All of my remaining comments are non-blocking, so approving.

File: cpp/src/io/utilities/hostdevice_vector.hpp [hunk 24]
```diff
@@ -190,7 +177,7 @@ class hostdevice_vector {
...
   T* host_data        = nullptr;
```
[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/utilities/hostdevice_vector.hpp:179]: Unresolved for the "remove `host_data`" part. I think we should make the host side consistent with device, now that we can.

File: cpp/include/cudf/io/config_utils.hpp
Comment (vuule) [inline cpp/include/cudf/io/config_utils.hpp:1]: IMO the file name should probably be `memory_resource.hpp` (or `host_memory_resource.hpp`), at least based on the similarity with `default_stream.hpp` header.

File: cpp/src/io/utilities/config_utils.cpp [hunk 25]
```diff
@@ -80,4 +88,38 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
...
+
+CUDF_EXPORT inline auto& host_mr()
+{
+  static rmm::host_async_resource_ref host_mr = default_pinned_mr();
+  return host_mr;
+}
+
+}  // namespace detail
+
+rmm::host_async_resource_ref set_host_memory_resource(rmm::host_async_resource_ref mr)
```
Comment (vuule) [inline cpp/src/io/utilities/config_utils.cpp:111]: Now I realize why the header mirrors this file's name :)
We should still probably pick a better name for the public header (and I can say "better", I named this file! :D ).

File: cpp/include/cudf/io/config_utils.hpp
Comment (harrism) [reply cpp/include/cudf/io/config_utils.hpp:1]: I missed this. Agree with @vuule

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp (hunk 23 above)
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:139]: This is a relic of the origins of this class: it came cudf::detail::pinned_host_vector, which came from some old experimental allocator in thrust itself.  Looks like thrust doesn't attempt to pass that parameter any more so I'll remove it.

File: cpp/include/cudf/detail/utilities/rmm_host_vector.hpp [hunk 26]
```diff
@@ -0,0 +1,207 @@
...
+   */
+  rmm_host_allocator(rmm::host_async_resource_ref _mr, rmm::cuda_stream_view _stream)
+    : mr(_mr), stream(_stream)
+  {
+  }
+
+  /**
+   * @brief Copy constructor
+   */
+  rmm_host_allocator(rmm_host_allocator const& other) = default;
```
[the remark this answers is not in the document]
Comment (nvdbaranec) [reply cpp/include/cudf/detail/utilities/rmm_host_vector.hpp:111]: I think this would be better done as a followup.  There are a number of places in cudf code using the assignment operator and thrust itself hits the copy constructor for mysterious reasons.  For example, just calling reserve on the wrapping `thrust::host_vector`  causes it to happen (`h_data.reserve(max_size);`).  Something happening internally in `thrust::detail::contiguous_storage`

File: cpp/src/io/utilities/config_utils.cpp (hunk 25 above)
Comment (nvdbaranec) [reply cpp/src/io/utilities/config_utils.cpp:111]: Any particular opinions?  It does cover a number of things beyond just the host allocator.

**Review by vuule — APPROVED** [review]
Thank you for addressing the feedback! Looks very clean now :fire:
